src/server/game/AI/CoreAI/UnitAI.h

```diff
@@ -152,94 +152,175 @@ class TC_GAME_API UnitAI
-            if (position >= threatlist.size())
+            ThreatManager& mgr = GetThreatManager();
+            // shortcut: if we ignore the first <offset> elements, and there are at most <offset> elements, then we ignore ALL elements
+            if (mgr.GetThreatListSize() <= offset)
                 return nullptr;
 
             std::list<Unit*> targetList;
-            Unit* currentVictim = nullptr;
-            if (auto currentVictimReference = GetThreatManager().getCurrentVictim())
+            if (targetType == SELECT_TARGET_MAXDISTANCE || targetType == SELECT_TARGET_MINDISTANCE)
             {
-                currentVictim = currentVictimReference->getTarget();
-
-                // Current victim always goes first
-                if (currentVictim && predicate(currentVictim))
+                for (ThreatReference* ref : mgr.GetUnsortedThreatList())
+                {
+                    if (ref->IsOffline())
+                        continue;
+                    targetList.push_back(ref->GetVictim());
+                }
+            }
+            else
+            {
+                Unit* currentVictim = mgr.GetCurrentVictim();
+                if (currentVictim)
                     targetList.push_back(currentVictim);
+                for (ThreatReference* ref : mgr.GetSortedThreatList())
+                {
+                    if (ref->IsOffline())
+                        continue;
+                    Unit* thisTarget = ref->GetVictim();
+                    if (thisTarget != currentVictim)
+                        targetList.push_back(thisTarget);
+                }
             }
 
-            for (HostileReference* hostileRef : threatlist)
+            // shortcut: the list certainly isn't gonna get any larger after this point
+            if (targetList.size() <= offset)
+                return nullptr;
+
+            // right now, list is unsorted for DISTANCE types - re-sort by MAXDISTANCE
+            if (targetType == SELECT_TARGET_MAXDISTANCE || targetType == SELECT_TARGET_MINDISTANCE)
+                SortByDistance(targetList, false);
+            // then reverse the sorting for MIN sortings
+            if (targetType == SELECT_TARGET_MINTHREAT || targetType == SELECT_TARGET_MAXDISTANCE)
+                targetList.reverse();
+            // now pop the first <offset> elements
+            while (offset)
             {
-                if (currentVictim != nullptr && hostileRef->getTarget() != currentVictim && predicate(hostileRef->getTarget()))
-                    targetList.push_back(hostileRef->getTarget());
-                else if (currentVictim == nullptr && predicate(hostileRef->getTarget()))
-                    targetList.push_back(hostileRef->getTarget());
+                targetList.pop_front();
+                --offset;
+            }
+            // and finally filter by predicate
+            for (auto it = targetList.begin(), end = targetList.end(); it != end;)
```

Comment:
my previous filter then offset comment also applies to single unit selection, not just list